src/quorum/majority.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+use super::{AckedIndexer, Index, VoteResult};
+use crate::{DefaultHashBuilder, HashSet};
+use std::mem::MaybeUninit;
+use std::{cmp, slice, u64};
+
+/// A set of IDs that uses majority quorums to make decisions.
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct Configuration {
+    pub(crate) voters: HashSet<u64>,
+}
+
+impl Configuration {
+    /// Creates a new configuration using the given IDs.
+    pub fn new(voters: HashSet<u64>) -> Configuration {
+        Configuration { voters }
+    }
+
+    /// Creates an empty configuration with given capacity.
+    pub fn with_capacity(cap: usize) -> Configuration {
+        Configuration {
+            voters: HashSet::with_capacity_and_hasher(cap, DefaultHashBuilder::default()),
+        }
+    }
+
+    /// Returns the MajorityConfig as a sorted slice.
+    pub fn slice(&self) -> Vec<u64> {
+        let mut voters: Vec<_> = self.voters.iter().cloned().collect();
+        voters.sort();
+        voters
+    }
+
+    /// Computes the committed index from those supplied via the
+    /// provided AckedIndexer (for the active config).
+    ///
+    /// The bool flag indicates whether the index is computed by group commit algorithm
+    /// successfully.
+    ///
+    /// Eg. If the matched indexes are [2,2,2,4,5], it will return 2.
+    /// If the matched indexes and groups are `[(1, 1), (2, 2), (3, 2)]`, it will return 1.
+    pub fn committed_index(&self, use_group_commit: bool, l: &impl AckedIndexer) -> (u64, bool) {
+        if self.voters.is_empty() {
+            // This plays well with joint quorums which, when one half is the zero
+            // MajorityConfig, should behave like the other half.
+            return (u64::MAX, false);
+        }
+
+        let mut stack_arr: [MaybeUninit<Index>; 7] = unsafe { MaybeUninit::uninit().assume_init() };
```

Comment:
I prefer `SmallVec` for this.

Reply:
I didn't use `SmallVec` on purpose. On the one hand, it's such a small piece of code that doesn't need to introduce another dependency; on the other hand, it has less condition branch as every operation is inlined and optimized when using array.